 # Analysing App Data From Playstore and Appstore
 The project is about exploring app data from Google playstore and Apple appstore to help our developers understand type of apps that are more likely to attract users.

In [37]:
from csv import reader

#for Google playstore
opened_file = open('C:\\Users\\USER\\Documents\\python\\googleplaystore.csv', encoding= "utf8")
read_file =reader(opened_file)
andriod = list(read_file)
andriod_header = andriod[0]
andriod = andriod[1:]

#for Apple store
opened_file = open('C:\\Users\\USER\\Documents\\python\\Applestore.csv', encoding= 'utf8')
read_file =reader(opened_file)
ios = list(read_file)
ios_header = ios[0]
ios = ios[1:]


function to explore both datasets 

In [38]:
def explore_data(dataset, start, end, rows_and_columns = False):
    dataset_slice = dataset[start:end]
    for row in dataset_slice:
        print(row)
        print('\n')
        
    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

In [39]:
print(ios_header)
print(explore_data(ios, 0, 2, True))

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']
['284882215', 'Facebook', '389879808', 'USD', '0', '2974676', '212', '3.5', '3.5', '95', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0', '2161558', '1289', '4.5', '4', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


Number of rows: 7197
Number of columns: 16
None


for the **ios** apps,
we have 7197 apps and 16 columns.
the columns that might be helpful are track_name, currency, rating_count_ver and prime_genre, price, rating_count_tot

In [40]:
print(andriod_header)
print(explore_data(andriod, 0, 2, True))

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']
['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


Number of rows: 10841
Number of columns: 13
None


for the andriod apps,
we have 10841 apps and 13 columns.
the columns that might  be useful are app, category, reviews, installs, type, price, genres.

In [41]:
print(andriod[7939])
print(andriod[10472])
print(andriod_header)

['Curriculum vitae App CV Builder Free Resume Maker', 'BUSINESS', '4.5', '4571', '3.9M', '500,000+', 'Free', '0', 'Everyone', 'Business', 'July 8, 2018', '3.3', '4.1 and up']
['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up', '']
['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


In [42]:
del andriod[10472]

In [43]:
len(andriod)

10840

the length of andriod apps have been reduced to 10840. 
this is because andriod[10472] is found with errors:
1. category gives 1.9 which cannot be correct.
2. review gives 19. the higest review for any app is '5'.
3. genres cannt be empty.

The Google play data has been discovered to contaibn duplicate entries (i.e it captures some apps data more than once). To confirm, I will print some of the duplicate entries below.

In [44]:
for app in andriod:
    name = app[0]
    if name == 'Facebook':
        print(app)

['Facebook', 'SOCIAL', '4.1', '78158306', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'August 3, 2018', 'Varies with device', 'Varies with device']
['Facebook', 'SOCIAL', '4.1', '78128208', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'August 3, 2018', 'Varies with device', 'Varies with device']


In [45]:
for app in andriod:
    name = app[0]
    if name == 'Instagram':
        print(app)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


In [46]:
duplicate_apps = []
unique_apps = []

for app in andriod:
    name = app[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)
        
print('Number of duplicate apps:', len(duplicate_apps), '\n')
print('Examples of duplicate apps are:', duplicate_apps[:20])

Number of duplicate apps: 1181 

Examples of duplicate apps are: ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software', 'MailChimp - Email, Marketing Automation', 'Crew - Free Messaging and Scheduling', 'Asana: organize team projects', 'Google Analytics', 'AdWords Express']


In [47]:
reviews_max = {}
for app in andriod:
    name = app[0]
    n_reviews = float(app[3])
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
    elif name not in reviews_max:
        reviews_max[name] = n_reviews
print(len(reviews_max))

9659


In [48]:
andriod_clean = []
already_added = []
for app in andriod:
    name = app[0]
    n_reviews = float(app[3])
    if n_reviews == reviews_max[name] and name not in already_added:
        andriod_clean.append(app)
        already_added.append(name)

In [49]:
print(andriod_clean[:3])
print('\n', already_added[:3])
print('\n', len(andriod_clean))
print('\n', len(already_added))

[['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'], ['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up'], ['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']]

 ['Photo Editor & Candy Camera & Grid & ScrapBook', 'U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'Sketch - Draw & Paint']

 9659

 9659


Next we are to remove all non English apps from our data (both and andriod and ios data). This is because our analysis should be directed towards an English speaking audience.

In [50]:
print(ord('b'))
print(ord('a'))
print(ord('A'))


98
97
65


In [51]:
def eng_fn(string):
    for name in string:
        if ord(name) > 127:
            return False
    return True
    
print(eng_fn('Instagram'))
print(eng_fn('Docs To Go'))
print(eng_fn('Instachat 😜'))

True
True
False


The function above checks a string and return true if the string name is English. But the function does not work properly yet, it does not recognise apps names with emojis and special characters.

In [52]:
def eng_fn(string):
    first = 0
    for name in string:
        if ord(name) > 127:
            first += 1
        if first > 3:
            return False
    return True
print(eng_fn('Instagram'))
print(eng_fn('Docs To Go'))
print(eng_fn('Instachat 😜')) 
print(ios[813][1])
print(eng_fn('爱奇艺PPS -《欢乐颂2》电视剧热'))

True
True
True
爱奇艺PPS -《欢乐颂2》电视剧热播
False


The modified function would remove any app name that have more than 3 non English letters. The function is still not perfect.

In [53]:
andriod_english =[]
ios_english = []
for app in andriod_clean:
    name = app[0]
    if eng_fn(name):
        andriod_english.append(app)
        

for app in ios:
    name = app[1]
    if eng_fn(name):
        ios_english.append(app)
        
print(explore_data(andriod_english, 0, 2, True))
print(explore_data(ios_english, 0, 2, True))

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Number of rows: 9614
Number of columns: 13
None
['284882215', 'Facebook', '389879808', 'USD', '0', '2974676', '212', '3.5', '3.5', '95', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0', '2161558', '1289', '4.5', '4', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


Number of rows: 6183
Number of columns: 16
None


In [54]:
andriod_free =[]
ios_free = []

for app in andriod_english:
    price = app[7]
    if price == '0':
        andriod_free.append(app)
        
for app in ios_english:
    price = app[4]
    if price == '0':
        ios_free.append(app)
        
print(explore_data(andriod_free, 0, 2, True))
print(explore_data(ios_free, 0, 2, True))

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Number of rows: 8864
Number of columns: 13
None
['284882215', 'Facebook', '389879808', 'USD', '0', '2974676', '212', '3.5', '3.5', '95', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0', '2161558', '1289', '4.5', '4', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


Number of rows: 3222
Number of columns: 16
None


The above code goes through both datas and isolate the free apps in a new list. This is our final data cleansing process. we have 8864 apps in our andriod and 3222 in our ios data left.

We are to develop an app that would be succcessful on both the App store and the Google Playstore. To do this, we need to find app profiles that are successful on both markets. Any app profile that works on both markets is more likely to attract both the Playstore and App store users, thereby making the app mpre successful. Our validation strategy is:
1. Build a minimal Andriod version of the app and add it to play store. This is because we have more app data from the Playstore than the App store.
2. If the app has a good response from users, develop it further. This is needed to keep existing users and also attract more new users.
3. If the app is profitable after 6 months, we build an IOS version of the app and add it to App store. This 6 months can be referred to as the probabtion period.

To generate a frequency table for the Google Play Store apps, we would need: App, Category, Genres.
For the ios Appstore, we would need: Track_name, prime_genre.

In [55]:
def freq_table(dataset, index):
    freq = {}
    total = 0
    for app in dataset:
        total += 1
        value = app[index]
        if value in freq:
            freq[value] += 1
        else:
            freq[value] = 1
    percent = {}
    for key in freq:
        percentage = (freq[key]/ total) * 100
        percent[key]= percentage
    return percent

def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)
    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])
        
print(display_table(andriod_free, 1))
print('\n')
print(display_table(andriod_free, 9))
print('\n')
print(display_table(ios_free,11))

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

For the ios App store:
- The most common genre is Games
- The 2nd most common genre is Entertainment.
- The general impression is that App Store for English apps is dominated by apps designed for fun (games, entertainment, sports, music, etc.) while practical apps (education, shopping, utilities, productivity, etc0 are more rare. However, the fact that fun apps are more doesn't imply that they also have the greatest number of users.

For the Google Play store:
- The most common genres are tools, followed by entertainment and education.
- The google data shows a balance landscape of both fun and practical apps

In [56]:
print(freq_table(ios_free, -5))

{'Social Networking': 3.2898820608317814, 'Photo & Video': 4.9658597144630665, 'Games': 58.16263190564867, 'Music': 2.0484171322160147, 'Reference': 0.5586592178770949, 'Health & Fitness': 2.0173805090006205, 'Weather': 0.8690254500310366, 'Utilities': 2.5139664804469275, 'Travel': 1.2414649286157666, 'Shopping': 2.60707635009311, 'News': 1.3345747982619491, 'Navigation': 0.186219739292365, 'Lifestyle': 1.5828677839851024, 'Entertainment': 7.883302296710118, 'Food & Drink': 0.8069522036002483, 'Sports': 2.1415270018621975, 'Book': 0.4345127250155183, 'Finance': 1.1173184357541899, 'Education': 3.662321539416512, 'Productivity': 1.7380509000620732, 'Business': 0.5276225946617008, 'Catalogs': 0.12414649286157665, 'Medical': 0.186219739292365}


In [57]:
for genre in freq_table(ios_free, -5):
    total = 0
    len_genre = 0
    for app in ios_free:
        genre_app = app[-5]
        if genre_app == genre:
            numb = float(app[5])
            total += numb
            len_genre += 1
    avg_user_rat = total / len_genre
    print(genre, ':', avg_user_rat)

Social Networking : 71548.34905660378
Photo & Video : 28441.54375
Games : 22788.6696905016
Music : 57326.530303030304
Reference : 74942.11111111111
Health & Fitness : 23298.015384615384
Weather : 52279.892857142855
Utilities : 18684.456790123455
Travel : 28243.8
Shopping : 26919.690476190477
News : 21248.023255813954
Navigation : 86090.33333333333
Lifestyle : 16485.764705882353
Entertainment : 14029.830708661417
Food & Drink : 33333.92307692308
Sports : 23008.898550724636
Book : 39758.5
Finance : 31467.944444444445
Education : 7003.983050847458
Productivity : 21028.410714285714
Business : 7491.117647058823
Catalogs : 4004.0
Medical : 612.0


In th code above, we calculated the average number of installs for each genre of the ios app data. We take the total number of user ratings as a proxy which can be find in the rating_count_tot column.

From the above findings, my app genre recommendation for the ios App store is Navigation, Reference and Business.

In [58]:
print(freq_table(andriod_free, 1))

{'ART_AND_DESIGN': 0.6430505415162455, 'AUTO_AND_VEHICLES': 0.9250902527075812, 'BEAUTY': 0.5979241877256317, 'BOOKS_AND_REFERENCE': 2.1435018050541514, 'BUSINESS': 4.591606498194946, 'COMICS': 0.6204873646209386, 'COMMUNICATION': 3.2378158844765346, 'DATING': 1.861462093862816, 'EDUCATION': 1.1620036101083033, 'ENTERTAINMENT': 0.9589350180505415, 'EVENTS': 0.7107400722021661, 'FINANCE': 3.7003610108303246, 'FOOD_AND_DRINK': 1.2409747292418771, 'HEALTH_AND_FITNESS': 3.0798736462093865, 'HOUSE_AND_HOME': 0.8235559566787004, 'LIBRARIES_AND_DEMO': 0.9363718411552346, 'LIFESTYLE': 3.9034296028880866, 'GAME': 9.724729241877256, 'FAMILY': 18.907942238267147, 'MEDICAL': 3.531137184115524, 'SOCIAL': 2.6624548736462095, 'SHOPPING': 2.2450361010830324, 'PHOTOGRAPHY': 2.944494584837545, 'SPORTS': 3.395758122743682, 'TRAVEL_AND_LOCAL': 2.33528880866426, 'TOOLS': 8.461191335740072, 'PERSONALIZATION': 3.3167870036101084, 'PRODUCTIVITY': 3.892148014440433, 'PARENTING': 0.6543321299638989, 'WEATHER': 

In [64]:
for category in freq_table(andriod_free, 1):
    total = 0
    len_category = 0
    for app in andriod_free:
        category_app = app[1]
        if category_app == category:
            install = app[5]
            install = install.replace('+', '')
            install= install.replace(',', '')
            total += float(install)
            len_category += 1
        avg_num_install = total / len_category
        print(category, ':', avg_num_install)

ART_AND_DESIGN : 10000.0
ART_AND_DESIGN : 2505000.0
ART_AND_DESIGN : 18336666.666666668
ART_AND_DESIGN : 13777500.0
ART_AND_DESIGN : 11032000.0
ART_AND_DESIGN : 9201666.666666666
ART_AND_DESIGN : 8030000.0
ART_AND_DESIGN : 7151250.0
ART_AND_DESIGN : 6357777.777777778
ART_AND_DESIGN : 5822000.0
ART_AND_DESIGN : 5383636.363636363
ART_AND_DESIGN : 5768333.333333333
ART_AND_DESIGN : 5332307.692307692
ART_AND_DESIGN : 4958571.428571428
ART_AND_DESIGN : 4628333.333333333
ART_AND_DESIGN : 4370312.5
ART_AND_DESIGN : 4113823.529411765
ART_AND_DESIGN : 4163055.5555555555
ART_AND_DESIGN : 4470263.157894737
ART_AND_DESIGN : 4251750.0
ART_AND_DESIGN : 4054047.619047619
ART_AND_DESIGN : 3892500.0
ART_AND_DESIGN : 3725434.782608696
ART_AND_DESIGN : 3570625.0
ART_AND_DESIGN : 3447800.0
ART_AND_DESIGN : 3319038.4615384615
ART_AND_DESIGN : 3196481.4814814813
ART_AND_DESIGN : 3085892.8571428573
ART_AND_DESIGN : 2982931.0344827585
ART_AND_DESIGN : 2885166.6666666665
ART_AND_DESIGN : 2795322.580645161
ART_

ZeroDivisionError: division by zero